Add in_parallel step

itsdalmo opened the pull request:
Closes #2628 by adding support for the following:

```yaml
jobs:
  - name: example
    plan:
    - get: every-minute
      trigger: true
    - in_parallel:
        limit: 1
        fail_fast: true
        steps:
        - task: echo
          config:
           ...
        - task: echo
          config:
            ...
```

InParallel ensures that `limit` steps are running at any one time and until all steps are complete. If `limit` is not specified, the default behavior is identical to an `aggregate` step and runs everything at once. Setting `limit` to 1 makes it behave like `do`. When cancelling a job, `in_parallel` will not schedule (`Run()`) the pending tasks and instead return the `ctx.Err()` as soon as possible. When setting `fail_fast`, `in_parallel` will abort all currently running tasks and refrain from scheduling pending tasks.

### Todo

- [ ] Revert changes to `examples/hello-world-every-minute.yml`.
- [x] Submit a PR to [concourse/docs](https://github.com/concourse/docs).
- [ ] Decide on YAML structure and naming (questions in [this comment](https://github.com/concourse/concourse/issues/2628#issuecomment-475908990)).

PS: Sorry for the big PR, but `parallel` is essentially a copy/paste of `aggregate` with very slight modifications. Was originally thinking I could refactor so both used the same underlying implementation, but I think that would only allow us to reuse the code in `exec` which does not have that big of an impact on the diff.

vito commented on atc/plan.go:
```diff
@@ -5,6 +5,7 @@ type Plan struct {
 	Attempts []int  `json:"attempts,omitempty"`
 
 	Aggregate *AggregatePlan `json:"aggregate,omitempty"`
+	Parallel  *ParallelPlan  `json:"parallel,omitempty"`
```
:thinking: The name feels slightly off; most of them are adjectives or verbs (except `task` I guess).

What do you think of `in_parallel`? I kinda like that as it fits alongside `on_success`/`do`/etc.

itsdalmo replied:
Makes sense, I've changed it to `in_parallel` 👍

vito commented on atc/config.go:
```diff
@@ -306,6 +306,11 @@ type PlanConfig struct {
+	// a nested chain of steps to run in parallel
+	Parallel *PlanSequence `yaml:"parallel,omitempty" json:"parallel,omitempty" mapstructure:"parallel"`
+	// limit paralelism for a Parallel plan
+	MaxInParallel int `yaml:"max_in_parallel,omitempty" json:"max_in_parallel,omitempty" mapstructure:"max_in_parallel"`
```
This name feels awkward but I'm not sure what else to suggest. Maybe it just sounds weird because I'm used to `max_in_flight`. I thought it might make sense to reuse the same name but @clarafu pointed it it might be confusing to have both set - someone could assume `max_in_flight: 3` at the job level is like setting it on all the steps in the plan. Maybe we should reserve 'in flight' terminology for builds, then? :thinking: If so this name is probably fine.

itsdalmo replied:
I see what you mean, and I agree that it's best to keep `max_in_flight` separate since it could get a bit confusing. Especially when you consider indentation errors in the YAML, which would produce unexpected results if the keyword is recognised in both the job and plan? 

However, right now we have `in_parallel` and `max_in_parallel` - not sure if I like it or not 🤔If I'm thinking outside the box, I would suggest something like this:

- Add a `PlanSequence` in `atc.Config` which we call `steps`.
- Keep `in_parallel` but change it to take a pointer to an `int`.
- Add `in_sequence` and have it take a pointer to a `string`.

We'd use the pointers so that the values would be `nil` if the key was not present in the map, so we can return an error from `validate` if `steps` is specified without either `in_sequence` or `in_parallel`. Then we could represent all three desired functionalities:

```yaml
jobs:
  - name: example
    plan:
    - in_parallel: 3
      steps:
      - task: echo
        config:
         ...
    - in_parallel: # <- We see a 0, and set the limit to the length of the steps. Same as aggregate.
      steps:
      - task: echo
        config:
         ...
    - in_sequence: # <- Equivalent to "do"?
      steps:
      - task: echo
        config:
         ...
```

Might work 🤔

vito replied:
I think the last two options there are a little awkward, since knowing how the YAML is parsed, those keys actually point to blank values, and Concourse's parser would have to be careful to notice that they're there at all (since they'd be set to either empty strings or nil - I don't remember).

The first option feels natural but forces you to pick a number in scenarios where you don't really care (unless we support the later two options).

I agree that `in_parallel` and `max_in_parallel` feel a little weird, but it at least allows the common case of "I just want things to run in parallel" to look simple:

```yaml
in_parallel:
- task: echo1
  file: config.yml
- task: echo2
  file: config.yml
- task: echo3
  file: config.yml
```

What if we allow two forms of `in_parallel`?

The first, simplest form which takes no configuration:

```yaml
in_parallel:
- task: echo1
  file: config.yml
- task: echo2
  file: config.yml
- task: echo3
  file: config.yml
```

...and in more complicated scenarios, it's instead an object with `limit` and `steps`:

```
in_parallel:
  limit: 3
  steps:
  - task: echo1
    file: config.yml
  - task: echo2
    file: config.yml
  - task: echo3
    file: config.yml
```

This could also be extended to support `fail_fast: true` (#165).

On the other hand I could see the repetition of `in_parallel` in `max_in_parallel` being a strength of the name, given that they *are* very closely tied together. :man_shrugging: 

Any preference?

itsdalmo replied:
I think in your example, `in_parallel` would only be able to cover the current `aggregate` and the proposed limited parallelisation, and in that case I would suggest we just stick with `in_parallel` and `max_in_parallel` - mainly because I suspect indentation might trip users up? 

We could still add `fail_fast` to this PR and perhaps throw an error if this key is present in a non-parallel context? However, it seems like the current `atc.Validate()` doesn't really do strict validation for the plan configuration, so perhaps it's fine to just ignore `fail_fast` if it's specified in another context? 🤔

itsdalmo commented:
> My take is that the coverage may be a bit excessive owing to aggregate's storied history (it's one of the oldest step types). It might make sense to replace some of the more verbose/brittle tests with testflight tests instead. Or maybe they're actually fine and not the brittle/verbose tests I'm remembering from aggregate. Like I said I haven't reviewed thoroughly yet. 🙂

You are right, it is the very same tests as in `aggregate` 😂I'm happy to go over the tests and rewrite them so that we only keep essential tests; there are a lot of tests to ensure that other steps/hook work within a `parallel`, which are probably safe to remove since each step is just executed with `step.Run(ctx, state)`.

I can leave a comment on the tests I think can be deleted.

itsdalmo commented on atc/scheduler/factory/factory_parallel_test.go:
```diff
@@ -0,0 +1,210 @@
+		})
+	})
+
+	Context("when I have nested parallel steps", func() {
```
I don't see how this can fail if the buildFactory can produce a plan from a `ParallelPlan`. I.e. the implementation is already identifying `Parallel` and recursing to produce a plan for substeps that may include another `parallel`. Delete?

vito replied:
Agreed

itsdalmo commented on atc/scheduler/factory/factory_parallel_test.go:
```diff
@@ -0,0 +1,210 @@
+		})
+	})
+
+	Context("when I have a parallel with hooks", func() {
```
Delete? If hooks work for any other step type, it should work for Parallel?

vito replied:
Agreed

itsdalmo commented on atc/scheduler/factory/factory_parallel_test.go:
```diff
@@ -0,0 +1,210 @@
+		})
+	})
+
+	Context("when I have a hook on a parallel", func() {
```
Delete? If hooks work for any other step type, it should work for Parallel?

vito replied:
Agreed

itsdalmo commented on atc/job_config_test.go:
```diff
@@ -502,6 +502,86 @@ var _ = Describe("JobConfig", func() {
 				})
 			})
 
+			Context("when a simple parallel plan is the first step", func() {
```
I think this test, and the ones following, are mostly testing `get` in relation to `PlanSequence` and not `aggregate` or `parallel` specifically. But if the desire is to move from `aggregate` to `parallel` in the future, then I think it makes sense to just keep them?

itsdalmo replied:
Same with the `exec_engine` tests. If the plan is to eventually move from `aggregate` to `parallel` then I think it makes sense to keep the tests; unless they can be cut completely?

vito replied:
I think these tests are just covering the recursing through all the different types of steps which contain other steps. It probably makes sense to keep it, since we need to recurse through the `Parallel` fields as well. This is necessary for finding all the 'job inputs', which are determined by looking through the entire plan for `get` steps.

jchesterpivotal commented:
I'd feel more comfortable if `aggregate` got an extra dial (`max_in_flight`, `throttle`, whatever), rather than having an entirely new step name that has overlapping functionality.

marco-m commented:
I agree with @jchesterpivotal here. Although the name `parallel` is more suitable than `aggregate`, having both names, besides code duplication, also adds user confusion, and I see already the same pipeline sporting both. +1 to add the max_in_flight to the existing aggregate.

pivotal-jwinters commented:
@vito I tend to agree with @jchesterpivotal as well. I know you wanted to migrate away from the `aggregate` step as per [this comment](https://github.com/concourse/concourse/issues/2628#issuecomment-472133539), but maybe we can just alias the name and deprecate the `aggregate` name over time?

itsdalmo commented:
The code changes are minimal, so all we currently need is a decision on:

- Add `parallel` step or extend `aggregate`?
- What should the config flag be called: `max_in_flight`, `max_in_parallel`, `limit` or `throttle`?

And keep in mind that if we extend `aggregate`, we have to add the above flag at the root of `atc.PlanConfig` (not nested inside a struct), less we will break existing pipelines. If we add `parallel` we are free to do whatever we want, at the cost of some code duplication and confusion (unless we are saying that `aggregate` is deprecated?).

I'm happy to implement this whatever way you guys want so as to get #2628 closed. Only thing I have to add is that `max_in_flight` might get "overloaded" if it is used in too many places.

vito commented:
@jchesterpivotal @marco-m @pivotal-jwinters I don't plan on supporting both `aggregate` and `in_parallel` forever - the point was to give users a reason to migrate to the new name by giving it more useful functionality (max-in-flight, fail-fast) and deprecate `aggregate` as `in_parallel` adoption grows.

Having an alias seems like it would be even more confusing - you've then shared the internal implementation, which users don't see, but duplicated *all* functionality, which users *do* see. Wouldn't it be clearer to have a deprecated, less-useful step and a newer step with a more meaningful name and more functionality?

marco-m commented:
@vito then I would suggest the following:
- introduce a check that returns error if the same pipeline has both `aggregate` and `in_parallel`. This check should be introduced in this PR or immediately after. The error message should mention that `aggregate` is deprecated.
- Deprecate `aggregate` immediately starting with 5.1. Add a check to `fly` that prints a deprecation warning and mentions _when_ aggregate will be removed.
- Remove `aggregate` in 6.0.

vito commented:
@marco-m Agreed. I'm not *too* sure about mixed `aggregate`/`in_parallel` being an error vs. a deprecation warning, but we at least have all the plumbing for emitting warnings.

@itsdalmo How's that sound?

Plumbing for deprecations:

https://github.com/concourse/concourse/blob/41f54f59330f2b6be55e05eb56166cb97494a64a/atc/validate.go#L27:17

->

https://github.com/concourse/concourse/blob/41f54f59330f2b6be55e05eb56166cb97494a64a/atc/api/configserver/save.go#L110

->

https://github.com/concourse/concourse/blob/41f54f59330f2b6be55e05eb56166cb97494a64a/fly/commands/internal/setpipelinehelpers/atc_config.go#L83

clarafu commented:
If we are trying to deprecate the `aggregate` step, then I would think that creating a new `parallel` step makes sense?

I feel like using the `max_in_flight` keyword again would be quite confusing especially if you have a very long pipeline config and you arn't sure if the `max_in_flight` is for the aggregate step or for the build plan, in which the keyword means two completely different things.

marco-m commented:
@clarafu I agree, `max_in_parallel` seems more appropriate.

itsdalmo commented:
@vito - I've added `fail_fast` in the latest commits. Struggled a bit with this one because it seems like `Run()` for the steps do not return an error when the task script fails - i.e. the only way I could get the expected behaviour was by using `s.Succeeded()`. Am I doing something wrong or is this working as intended?

In the current commit, `in_parallel` with `fail_fast: true` will interrupt currently running steps, stop scheduling pending tasks. This is the behaviour I would expect myself, so that is what I implemented, but perhaps you had something else in mind?

Also, I'm wondering if perhaps the configuration should _only_ be your second suggestion:

```yaml
in_parallel:
  limit: 3
  steps:
  - task: echo1
    file: config.yml
  - task: echo2
    file: config.yml
  - task: echo3
    file: config.yml
  fail_fast: true
```

Seems cleaner when related configuration goes in a map like this? What do you think?

itsdalmo commented:
@vito I added support for the shorthand form of `in_parallel` in the latest commit. I have not added custom serialization for YAML/JSON, so `get-pipeline` and `set-pipeline` will always return/diff `in_parallel: [ ... ]` as `in_parallel: { steps: [ ... ] }`. Returning the long form of the config feels clean to me since e.g. YAML anchors will be expanded on `get` and `set-pipeline` also - but let me know if you want it changed. Also added a deprecation warning if `aggregate` is set, as part of the validation.

Thoughts?

itsdalmo commented:
Thoughts on this @vito? Are we approaching mergeable? 😅

vito commented:
@itsdalmo Sorry! I was focusing on getting 5.1 out the door, which we were on the verge of for a while, but it stretched on for longer than expected. I'll take another look at this today and merge if it's good to go. :+1:
